tanchan/_internal/localisation.py

```diff
@@ -0,0 +1,155 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from __future__ import annotations
+
+__all__: list[str] = []
+
+import typing
+
+import hikari
+import tanjun
+
+if typing.TYPE_CHECKING:
+    from collections import abc as collections
+
+
+# TODO: expand tanjun's special casing of * to cover other command fields like responses here
+_CommandTypes = typing.Literal["message", "message_menu", "slash", "user_menu"]
+_TYPE_TO_STR: dict[typing.Optional[hikari.CommandType], _CommandTypes] = {
+    hikari.CommandType.MESSAGE: "message_menu",
+    hikari.CommandType.SLASH: "slash",
+    hikari.CommandType.USER: "user_menu",
+    None: "message",
+}
+_FieldTypes = typing.Literal["help.category", "help.name", "help.description"]
+
+
+class MaybeLocalised:
+    """Class used for handling name and description localisation."""
+
+    __slots__ = ("default_value", "id", "_localise_id", "localised_values")
+
+    def __init__(
+        self,
+        field_type: _FieldTypes,
+        field: typing.Union[str, collections.Mapping[str, str], collections.Iterable[tuple[str, str]]],
+        /,
+        *,
+        cmd_type: typing.Union[hikari.CommandType, None] = None,
+        name: typing.Optional[str] = None,
+    ) -> None:
+        """Initialise an instance of MaybeLocalised.
+
+        Parameters
+        ----------
+        field_name
+            Name of the field being localised.
+
+            This is used in raised exceptions.
+        field
+            The string value(s) to use for this value.
+
+            If a [str][] is passed here then this will be used as the default
+            value and the field's id for overloading it with the localiser.
+
+            When a mapping is passed here, this should be a mapping of locales
+            to values. If an "id" fieldis included then this will be used as the
+            id for overloading it with the localiser and the first real value
+            will be used as the default value.
+
+        Raises
+        ------
+        RuntimeError
+            If no default value is provided when `filed` is a mapping.
+        """
+        if field_type == "help.category":
+            if name is not None:
+                raise RuntimeError("`name` cannot be passed for help.category fields")
+
+            self._localise_id = f"*:*:help.category:{name}"
```

Comment:
TODO: tanjun needs to also accept using the raw "*:*" string for getting localisations